File: android_demos_downloader.rb
# encoding: utf-8

# Install
#    this is a ruby program, so install ruby first
#    after ruby installed, install needed gems
#    gem install nokogiri active_support
#
# Usage
#    ruby android_demos_downloader.rb

require 'nokogiri'
require 'open-uri'
require 'active_support'
require 'pathname'
require "fileutils"

class DownLoader

  def start
    @file_list = {}
    @stack = []
    # @images = {}
    parse_index_page
    current_path = Pathname.new(File.dirname(__FILE__)).realpath.to_s + '/'
    save_codes(@file_list, current_path)
  end

  def parse_index_page
    file_name = 'file_list_json'
    if File.exists?(file_name)
      puts "read file list from file #{file_name}"
      json = ''
      File.open(file_name) do |file|
        file.each do |line|
          json << line
        end
      end
      @file_list = ActiveSupport::JSON.decode(json)
    else
      doc = Nokogiri::HTML(open('http://developer.android.com/intl/zh-cn/samples/index.html'))
      doc.css('#nav > .nav-section').each do |section|
        parse_file_list(section, @file_list)
      end
      json = ActiveSupport::JSON.encode(@file_list)
      file = File.new(file_name, 'w')
      file.puts json
      file.close
      puts "saved file list to file #{file_name}"
    end
  end

  def parse_file_list(section, list)
    links = section.css('> .nav-section-header > a')
    if links && links.length > 0
      attr = links[0].attr('title')
      if attr
        title = attr.class.to_s == 'String' ? attr : attr.value
        list[title] = {}
        index = 0
        section.children.css('> li').each do |li|
          index += 1
          parse_file_list(li, list[title])
        end
      end
    else
      link = section.css('> a')
      if link.length > 0
        url = link.attr('href').value
        file_name = link.attr('title').value
        if is_image?(file_name)
          arr = url.split('/')
          arr[-1] = file_name
          url = arr.join('/')
        end
        list[file_name] = url
      end
    end
  end

  def save_codes(file_list, current_path)
    file_list.each do |name, value|
      if value.class.to_s == 'String'
        unless File.exists? "#{current_path}#{name}"
          code = parse_code_page(name, value)
          puts "save: #{current_path}#{name}"
          save_code(current_path, name, code)
        end
      else
        dir_path = current_path + "#{name.gsub(/\./, '/')}/"
        save_codes(file_list[name], dir_path)
      end
    end
  end

  def parse_code_page(name, url)
    code = ''
    if is_image?(name)
      code = "http://developer.android.com#{url.start_with?('/') ? '' : '/'}#{url}"
    else
      puts "\nparse #{url}"
      t1 = Time.now
      page = open("http://developer.android.com#{url}")
      t2 = Time.now
      puts "download takes #{t2 - t1}s"
      doc = Nokogiri::HTML(page)
      wrappers = doc.css('#codesample-wrapper')
      if wrappers && wrappers[0]
        arr = []
        wrappers[0].content.split("\n").each do |line|
          unless /^\d+$/.match(line.strip) || line.strip == ''
            arr << line
          end
        end
        code = arr.join("\n")
      end
      t3 = Time.now
      puts "parse page takes #{t3 - t2}s"
      puts "total #{t3 - t1}s"
    end

    code
  end

  def save_code(path, name, code)
    mkdir(path)
    if is_image?(name)
      arr = code.split('/')
      full_name = "#{arr[-2]}/#{arr[-1]}"
      # if @images.include?(full_name)
      #   FileUtils.copy(@images[full_name], "#{path}#{name}")
      # else
      data=open(code) { |f| f.read }
      open("#{path}#{name}", "wb") { |f| f.write(data) }
      # @images[full_name] = "#{path}#{name}"
      # end
    else
      unless File.exists? "#{path}#{name}"
        file = File.new("#{path}#{name}", 'w')
        file.puts(code)
        file.close
      end
    end
  end

  def mkdir(path)
    current_path = Pathname.new(File.dirname(__FILE__)).realpath.to_s
    sub_path = path.gsub(current_path, '')
    dir_path = ''
    sub_path.split('/').each do |dir_name|
      if dir_name != ''
        dir_path << dir_name
        Dir.mkdir dir_path unless File.exists? dir_path
        dir_path << '/'
      end
    end
  end


  def is_image?(url)
    ['png', 'jpg', 'gif'].include?(url.split('.')[-1])
  end


end

downloader = DownLoader.new
downloader.start

File: android_demos_downloader.js
/**
 * Install
 *   before run this program, you should install needed libraries
 *   this is a node.js program, so install node.js first ( see http://nodejs.org/)
 *   after that, run command blow to install some node.js modules
 *   npm install jsdom jquery
 *
 * Usage
 *   node android_demos_downloader.js
 */

var fs = require('fs'),
    http = require('http'),
    jsdom = require('jsdom').jsdom,
    window = jsdom().createWindow(),
    path = require('path'),
    $ = require('jquery')(window);

var fileList = {},
//    images = {},
    stack = [],
    jsonFileName = 'file_list_json',
    webUrl = 'http://developer.android.com';


function getFileLists(callback) {
    fs.exists(jsonFileName, function (exists) {
        if (exists) {
            console.log('read file list from file: ' + jsonFileName);
            fs.readFile(jsonFileName, 'utf8', function (err, data) {
                if (err) {
                    throw err;
                }
                fileList = JSON.parse(data);
                callback();
            });
        } else {
            parseIndexPage(callback);
        }
    })
}

function parseIndexPage(callback) {
    var indexPageUrl = webUrl + '/samples/index.html';
    console.log('parse file list from ' + indexPageUrl);
    http.get(indexPageUrl, function (res) {
        var html = '';
        res.on('data', function (chunk) {
            html += chunk;
        });
        res.on('end', function () {
            $(html).find('#nav').children('.nav-section').each(function (_, li) {
                parseFileLists($(li), fileList);
            });
            //写入文件
            console.log('saving file list to file: ' + jsonFileName);
            var text = JSON.stringify(fileList);
            fs.writeFile(jsonFileName, text, function (err) {
                if (err) throw err;
                console.log('saved'); //文件被保存
                callback();
            });
        });
    });
}


function parseFileLists($section, files) {
    var title = $section.children('.nav-section-header').children('a').attr('title');
    if (title) {
        files[title] = {};
        $section.children('ul').children('li').each(function (_, li) {
            parseFileLists($(li), files[title]);
        });
    } else {
        var $a = $section.children('a');
        if ($a.length > 0) {
            var url = $a.attr('href');
            var fileName = $a.attr('title');
            if (isImage(fileName)) {
                var arr = url.split('/');
                arr[arr.length - 1] = fileName;
                url = arr.join('/');
            }
            files[fileName] = url;
        }
    }
}

var index = -1;

function saveCodes() {
    console.log('begin to download');
    saveNext();
}

function saveNext() {
    index++;
    var option = stack[index];
    if (option) {
        var path = option.path,
            name = option.name,
            url = option.url;

        parseAndSave(path, name, url);
    }
}

function createStack(list, path) {
    for (var name in list) {
        if (list.hasOwnProperty(name)) {
            var value = list[name];
            if (typeof value == 'string') {
                //parseAndSave(path, name, value);
                stack.push({
                    path: path,
                    name: name,
                    url: value
                });
            } else {
                var dir_path = path + name.split('.').join('/') + '/';
                createStack(value, dir_path);
            }
        }
    }
}

function parseAndSave(path, name, url) {
    fs.exists(path + name, function (exists) {
        if (!exists) {
            console.log('');
            parseCodePage(path, name, url, function (file_path, file_name, code) {
                saveCode(file_path, file_name, code);
            });
        } else {
            saveNext();
        }
    });
}

function parseCodePage(path, name, url, callback) {
    var code = '', arr;
    if (isImage(name)) {
        code = 'http://developer.android.com' + (url[0] == '/' ? '' : '/') + url;
        callback(path, name, code);
    } else {
        var pageUrl = webUrl + url;
        console.log('loading page ' + pageUrl);
        http.get(pageUrl, function (res) {
            var html = '';
            res.on('data', function (chunk) {
                html += chunk;
            });
            res.on('end', function () {
                console.log('parsing...');
                var content = $(html).find('#codesample-wrapper').text();
                arr = content.split('\n');
                var codeArr = [];
                for (var i = 0; i < arr.length; i++) {
                    var line = arr[i], text = $.trim(line);
                    if (!/^\d+$/.test(text) && text != '') {
                        codeArr.push(arr[i]);
                    }
                }
                var code = codeArr.join('\n');
                callback(path, name, code);
            });
        });
    }
}

function saveCode(path, name, code) {
    mkdirs(path, 0777, function () {
        if (isImage(name)) {
            var arr = code.split('/');
            var imgName = arr[arr.length - 2] + '/' + arr[arr.length - 1];
//            if (images[imgName]) {
//                copy(images[imgName], path + name);
//            } else {
                downloadImage(path + name, code);
//            }
        } else {
            fs.exists(path + name, function (exists) {
                if (!exists) {
                    console.log('saving ' + path + name);
                    fs.writeFile(path + name, code, function (err) {
                        if (err) throw err;
                        console.log('saved'); //文件被保存
                        saveNext();
                    });
                } else {
                    saveNext();
                }
            })
        }
    });
}

function copy(from, to) {
    console.log('copying file to ' + to);
    var fileReadStream = fs.createReadStream(from);
    var fileWriteStream = fs.createWriteStream(to);
    fileReadStream.pipe(fileWriteStream);

    fileWriteStream.on('close', function () {
        console.log('copy over');
        saveNext();
    });
}

function downloadImage(path, url) {
    fs.exists(path, function (exists) {
        if (!exists) {
            console.log('downloading image from ' + url);
            http.get(url, function (res) {
                var body = '';
                res.setEncoding('binary');
                res.on('data', function (chunk) {
                    body += chunk;
                });
                res.on('end', function () {
                    console.log('saving to ' + path);
                    fs.writeFile(path, body, 'binary', function (err) {
                        if (err) {
                            throw err;
                        }
                        console.log('saved');
                        var arr = url.split('/');
                        var imgName = arr[arr.length - 2] + '/' + arr[arr.length - 1];
//                        images[imgName] = path;
                        saveNext();
                    });
                });
            })
        } else {
            saveNext();
        }
    });
}

function isImage(name) {
    var arr = name.split('.');
    return ['png', 'jpg', 'git'].indexOf(arr[arr.length - 1]) > -1;
}

// 创建所有目录
function mkdirs(dirpath, mode, callback) {
    fs.exists(dirpath, function (exists) {
        if (exists) {
            callback(dirpath);
        } else {
            //尝试创建父目录，然后再创建当前目录
            mkdirs(path.dirname(dirpath), mode, function () {
                console.log("mkdir " + dirpath);
                fs.mkdir(dirpath, mode, callback);
            });
        }
    });
}

function start() {
    getFileLists(function () {
        console.log('create downloading list...');
        createStack(fileList, './');
        saveCodes();
    });
}

start();

// test
//downloadImage('./ic_launcher.png', 'http://developer.android.com/samples/BasicSyncAdapter/res/drawable-hdpi/ic_launcher.png');